### PR TITLE
[GLUTEN-3934][CH]Bug fix log function diff

### PR DESCRIPTION
**Pull request by KevinyhZou**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#3934)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

TEST BY UT

### CODE REVIEW

File: cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+        const auto * arg_node = parsed_args[0];
+        const std::string function_name = getName();
+        const auto * log_node = toFunctionNode(actions_dag, function_name, {arg_node});
+
+        auto result_type = log_node->result_type;
+        auto nullable_result_type = makeNullable(result_type);
+
+        const auto * null_const_node = addColumnToActionsDAG(actions_dag, nullable_result_type, Field());
+        const auto * nullable_log_node = ActionsDAGUtil::convertNodeType(actions_dag, log_node, nullable_result_type->getName(), log_node->result_name);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h:65]: 没必要强转。if函数会处理。

File: cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                null
+            else
+                log(c)
+        */
+        auto parsed_args = parseFunctionArguments(substrait_func, "", actions_dag);
+        if (parsed_args.size() != 1)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} requires exactly one arguments", getName());
+
+        const auto * arg_node = parsed_args[0];
+        const std::string function_name = getName();
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h:58]: getName()返回的是substrait function name, 不是ch function name。这里不能混为一谈

File: cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        const auto * log_node = toFunctionNode(actions_dag, function_name, {arg_node});
+
+        auto result_type = log_node->result_type;
+        auto nullable_result_type = makeNullable(result_type);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h:62]: 可以合并了，auto nullable_result_type = makeNullable(log_node->result_type);

File: cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+        const auto * null_const_node = addColumnToActionsDAG(actions_dag, nullable_result_type, Field());
+        const auto * nullable_log_node = ActionsDAGUtil::convertNodeType(actions_dag, log_node, nullable_result_type->getName(), log_node->result_name);
+        const DB::Float64 lowerBound = getParameterLowerBoundValue();
+        const auto * le_node = toFunctionNode(actions_dag, "lessOrEquals", {arg_node, addColumnToActionsDAG(actions_dag, result_type, lowerBound)});
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h:67]: lowerBound类型错了，应该是 arg_node的类型。

File: cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    const ActionsDAG::Node * parse(
+        const substrait::Expression_ScalarFunction & substrait_func,
+        ActionsDAGPtr & actions_dag) const override
+    {
+        /*
+            parse log(x) as
+            if (x <= c)
+                null
+            else
+                log(c)
```
Comment (exmy) [inline cpp-ch/local-engine/Parser/scalar_function_parser/logarithm.h:51]: 这里应该是 x <= 0,  log(x) ?

**Comment by KevinyhZou** [comment]
Fixed done.

**Review by liuneng1994 — APPROVED** [review]
LGTM

**Pull request merged**